src/app.ts

```diff
@@ -24,12 +24,72 @@ import ServicesController from "./controllers/servicesController";
 
 import { Express } from "express";
 import expressEnforcesSsl = require("express-enforces-ssl");
+import { Either } from "fp-ts/lib/Either";
 import {
   EnvironmentNodeEnv,
   EnvironmentNodeEnvEnum
 } from "./types/environment";
 import { User } from "./types/user";
 
+/**
+ * Return a response to the client.
+ */
+function respond(
+  response: Either<Error, IResponse>,
+  res: express.Response
+): void {
+  response.fold(
+    err => {
+      res.status(500).json({ message: err.message });
+    },
+    data => {
+      if (data.status === 301) {
+        res.redirect(data.body);
+      } else {
+        res.json(data.body);
+      }
+    }
+  );
+}
+
+/**
+ * Catch SPID authentication errors and redirect the client to
+ * clientErrorRedirectionUrl.
+ */
+const withSpidAuth = (
+  controller: AuthenticationController,
+  clientErrorRedirectionUrl: string,
+  clientLoginRedirectionUrl: string
+): ((
+  req: express.Request,
+  res: express.Response,
+  next: express.NextFunction
+) => void) => {
+  return (
+    req: express.Request,
+    res: express.Response,
+    next: express.NextFunction
+  ) => {
+    passport.authenticate("spid", async (err, user: User) => {
```

Comment:
authenticate definition is:

`authenticate(strategy: string | string[], callback?: (...args: any[]) => any): AuthenticateRet;`

so I don't think err has a specific type, or I'm wrong?

Reply:
yes, all cb params are `any` - so how do you know the 2nd is a `User`?

Reply:
in effect also the `User` type is incorrect, at this point the user is still `any`, it will be validated later in `AuthenticationController#acs`. Should I set `any` to both err and user here?`

Reply:
yes, I think you can just avoid specifying a type, it will be `any` from the function signature